### PR TITLE
Budget App

### PR DESCRIPTION
**Pull request by JacintaNgwu**
### Project requirements
**Design**
You should follow these [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
Colors.
Typography: font face, size and weight.
Layout: composition and space between elements.
NOTE: In these design guidelines there are several UIs that you won't need for this exercise; also, some pages are not given a design and you will create them following the design guidelines of the other pages.

Original design idea by [Gregoire Vella on Behance](https://www.behance.net/gregoirevella).

The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) requires that you give appropriate credit to the author. Therefore, you must do it in the README of your project.

**Interactions**

- Splash screen
- [ ] A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.

- Sign up and log in pages
- [ ] The user should be able to register in the app with full name, email and password (all mandatory).
- [ ] The user can log into the app using email and password.
- [ ] If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

- Home page (categories page)
- [ ] When the user logs in, they are presented with the categories page.
- [ ] For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
- [ ] When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
- [ ] There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

- Transactions page
- [ ] For a given category, the list of transactions is presented, ordered by the most recent.
- [ ] At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
- [ ] There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
- [ ] When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new category" page
- [ ] The user fills out a form to create a new category, indicating their name and icon (both mandatory).
- [ ] The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
- [ ] When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new transaction" page
- [ ] The user fills out a form to create a new transaction with:
1. name (mandatory)
2. amount (mandatory)
3. categories (mandatory at least one)
- [ ] The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
- [ ] When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

**Testing requirements**
- Create unit and integration tests for all the most important components of your RoR application.

**Technical requirements**
- You should use Postgres as your database.
- You should use devise for authentication.
- You should validate all user input to make sure that anyone with bad intentions cannot compromise your app.
- You can use a view template engine of your choice (.erb, .slim, .haml).
- The project should be deployed and accessible online.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -93,196 +53,134 @@ After you're finished please remove all the comments and instructions!
...
 <!-- LIVE DEMO -->
 
 ## 🚀 Live Demo <a name="live-demo"></a>
 
 > Add a link to your deployed project.
 
-- [Live Demo Link](https://google.com)
+- [Live Demo Link](https://budget-app-56iz.onrender.com)
+
+- [Video Presentation](https://www.loom.com/share/b7e5378aa77a4fd584c551e77cedf6a3)
```
Comment (BenMukebo) [inline README.md:76]: - [ ] Great job deploying your app to render however, I get this page when I try to access the registration or login page 👇🏻 . please, let's make sure our app is properly deployed and all functions are working as expected.

image

File: app/controllers/categories_controller.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+  def create
+    @author = current_user
+    @category = Category.new(category_params)
+    @category.author = @author
+
+    respond_to do |format|
+      if @category.save
+        format.html { redirect_to @category, notice: 'Category was successfully created.' }
+        format.json { render :show, status: :created, location: @category }
```
Comment (BenMukebo) [inline app/controllers/categories_controller.rb]: ###  "Add a new category or Group" page

- [x] According to the project requirements, When the user clicks (or taps) the "Save" button to create the new category, it must take the user to the categories Page on success and not the `category/:id` page. Kindly correct this as well. ✍🏻

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+  validates :name, presence: true, allow_blank: false
+  has_many :expenses, dependent: :destroy, foreign_key: :author_id
+  has_many :categories, dependent: :destroy, foreign_key: :author_id
```
Comment (BenMukebo) [inline app/models/user.rb]: - [x] Great job adding user name validation 👏🏻 Let's now validate all user input (email, password, ...) to make sure that anyone with bad intentions cannot compromise their app.

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+require 'rails_helper'
```
Comment (BenMukebo) [inline spec/models/user_spec.rb:1]: - [ ] Excellent job writing unit tests for most of the most important components of the RoR application. 👍🏻 
 Now, let's make sure to also add integration tests with Capybara for all the most important components of the RoR application. ✍🏻

image

File: app/views/categories/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        <%= render "shared/messages" %>
+        
+            <% @categories.each do |category| %>
+                <div class="c-index-cont__category__name">
+                <%= link_to category_path(category), class: "c-index-cont__category" do %>
+                        <div class="img-name">
+                        <%= image_tag(category.icon_url, class: "c-index-cont__category__name__icon") %>
+                        <p class="c-index-cont__category__name__text"><%= category.name %></p>
+                        </div>
+                        <div class="amt">
```
Comment (BenMukebo) [inline app/views/categories/index.html.erb]: - [OPTIONAL] The card should have a category image, name, `created date`, and amount. kindly let display the creation date for each category card. 🛀🏻 

Your Implementation | Original Design
:---:|:---:
image| image

**Review by BenMukebo — CHANGES_REQUESTED** [review]
## Hi @JacintaNgwu, 

You did a great job so far 👏🏼 👏🏼
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>


### To Highlight 🎉
- Professional README file. :heavy_check_mark:
- PR has a good title and summary. ✔️
- Good presentation video. ✔️ 
- The database schema reflect the ERD diagram. ✔️ 
- Associations between models has been setup in a correct way. ✔️
- Well set up devise for authentication. 🥇 

## STATUS: CHANGES REQUESTED  ♻️ 
_Please Check the comments under the review and make all required changes_
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers :clinking_glasses:  and Happy coding!!! :dancers:💻  

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Ismailco — APPROVED** [review]
### Hi @JacintaNgwu, 

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 

cr-approve

## Highlights

✔️ No linter errors
✔️ Used Gitflow
✔️ Work documented professionally

Good luck! 🎉 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**Pull request merged**